src/library_openal.js

```diff
@@ -1,1622 +1,4591 @@
-              AL.setSourceState(src, 0x1012 /* AL_PLAYING */);
-            } else {
-              AL.setSourceState(src, 0x1014 /* AL_STOPPED */);
+          var duration = 0.0;
+          // If the source is a looping static buffer, use native looping for gapless playback
+          if (src.type === 0x1028 /* AL_STATIC */ && src.looping) {
+            duration = Number.POSITIVE_INFINITY;
+            audioSrc.loop = true;
+            if (buf.audioBuf._loopStart) {
+              audioSrc.loopStart = buf.audioBuf._loopStart;
+            }
+            if (buf.audioBuf._loopEnd) {
+              audioSrc.loopEnd = buf.audioBuf._loopEnd;
             }
+          } else {
+            duration = (buf.audioBuf.duration - startOffset) / src.playbackRate;
           }
-        }
-        // Process all buffers that'll be played before the next tick.
-        else if (startOffset < (AL.QUEUE_LOOKAHEAD / 1000) && !entry.src) {
-          // If the start offset is negative, we need to offset the actual buffer.
-          var offset = Math.abs(Math.min(startOffset, 0));
-
-          entry.src = src.context.ctx.createBufferSource();
-          entry.src.buffer = entry.buffer;
-          entry.src.connect(src.gain);
-          if (src.playbackRate != 1.0) entry.src.playbackRate.value = src.playbackRate;
-          entry.src.duration = entry.buffer.duration / src.playbackRate;
-          if (typeof(entry.src.start) !== 'undefined') {
-            entry.src.start(startTime, offset);
-          } else if (typeof(entry.src.noteOn) !== 'undefined') {
-            entry.src.noteOn(startTime);
-#if OPENAL_DEBUG
-            if (offset > 0) {
-              Runtime.warnOnce('The current browser does not support AudioBufferSourceNode.start(when, offset); method, so cannot play back audio with an offset '+offset+' secs! Audio glitches will occur!');
+
+          audioSrc._startOffset = startOffset;
+          audioSrc._duration = duration;
+          audioSrc._skipCount = skipCount;
+          skipCount = 0;
+
+          audioSrc.connect(src.gain);
+
+          if (typeof(audioSrc.start) !== "undefined") {
+            // Sample the current time as late as possible to mitigate drift
+            startTime = Math.max(startTime, src.context.audioCtx.currentTime);
+            audioSrc.start(startTime, startOffset);
+          } else if (typeof(audioSrc.noteOn) !== "undefined") {
+            startTime = Math.max(startTime, src.context.audioCtx.currentTime);
+            audioSrc.noteOn(startTime);
+#if OPENAL_DEBUG
+            if (offset > 0.0) {
+              Runtime.warnOnce("The current browser does not support AudioBufferSourceNode.start(when, offset); method, so cannot play back audio with an offset "+startOffset+" secs! Audio glitches will occur!");
             }
 #endif
           }
 #if OPENAL_DEBUG
           else {
-            Runtime.warnOnce('Unable to start AudioBufferSourceNode playback! Not supported by the browser?');
```

Comment:
This PR changes all instances of strings from single quotes `'` to double quotes `"`. The convention in this repository in JS code is to use single quotes - is there a specific motivation for this change?